vnext/Microsoft.ReactNative.Cxx/ReactNotificationService.h

```diff
@@ -0,0 +1,298 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+#ifndef MICROSOFT_REACTNATIVE_REACTNOTIFICATIONSERVICE
+#define MICROSOFT_REACTNATIVE_REACTNOTIFICATIONSERVICE
+
+#include "ReactDispatcher.h"
+#include "ReactPropertyBag.h"
+
+namespace winrt::Microsoft::ReactNative {
+
+// Encapsulates the IReactPropertyName and the notification data type
+template <class T>
+struct ReactNotificationId : ReactPropertyName {
+  using NotificationDataType = T;
+  using ReactPropertyName::ReactPropertyName;
+
+  ReactPropertyName const &NotificationName() const noexcept {
+    return *this;
+  }
+};
+
+struct ReactNotificationSubscription {
+  ReactNotificationSubscription(std::nullptr_t = nullptr) noexcept {}
+
+  explicit ReactNotificationSubscription(IReactNotificationSubscription const &handle) noexcept : m_handle{handle} {}
+
+  IReactNotificationSubscription const &Handle() const noexcept {
+    return m_handle;
+  }
+
+  explicit operator bool() const noexcept {
+    return static_cast<bool>(m_handle);
+  }
+
+  ReactDispatcher Dispatcher() const noexcept {
+    return ReactDispatcher{m_handle ? m_handle.Dispatcher() : nullptr};
+  }
+
+  // Name of the notification.
+  ReactPropertyName NotificationName() const noexcept {
+    return ReactPropertyName{m_handle ? m_handle.NotificationName() : nullptr};
+  };
+
+  // True if the subscription is still active.
+  // This property is checked before notification handler is invoked.
+  bool IsSubscribed() const noexcept {
+    return m_handle ? m_handle.IsSubscribed() : false;
+  };
+
+  // Remove the subscription.
+  // Because of the multi-threaded nature of the notifications, the handler can be still called
+  // after the Unsubscribe method called if the IsSubscribed property is already checked.
+  // Consider calling the Unsubscribe method and the handler in the same IReactDispatcher
+  // to ensure that no handler is invoked after the Unsubscribe method call.
+  void Unsubscribe() const noexcept {
+    if (m_handle) {
```

Comment:
>if (m_handle) { [](start = 4, length = 15)

Why would we ever allow null handle? Wouldn't it be easier if m_handle is always an object?

Reply:
See the other answer and the constructor that accepts nullptr.

---
In reply to: [427626890](https://github.com/microsoft/react-native-windows/pull/4953#discussion_r427626890) [](ancestors = 427626890)